src/library/scala/concurrent/Future.scala

```diff
@@ -413,8 +413,37 @@ trait Future[+T] extends Awaitable[T] {
    * @return        a `Future` with the result of the application of `f` to the results of `this` and `that`
    * @group Transformations
    */
-  def zipWith[U, R](that: Future[U])(f: (T, U) => R)(implicit executor: ExecutionContext): Future[R] =
-    flatMap(r1 => that.map(r2 => f(r1, r2)))(if (executor.isInstanceOf[BatchingExecutor]) executor else parasitic)
+  def zipWith[U, R](that: Future[U])(f: (T, U) => R)(implicit executor: ExecutionContext): Future[R] = {
+    val sentinel = new Object()
+    val buffer = new AtomicReference[AnyRef](sentinel)
+    val outPromise = Promise[R]()
+
+    val myExecutor = if (executor.isInstanceOf[BatchingExecutor]) executor else parasitic
```

Comment:
@lihaoyi This is wrong—it would make `f` evaluated on the current thread if `this` or `that` has already been completed when the callback is added. It *must* be executed on the passed in EC. If you look at the default implementation of `zipWith` (which this PR is replacing) you can see that the choice of doing `if (executor.isInstanceOf[BatchingExecutor]) executor else parasitic` only applies to the `flatMap` and not the `map`: https://github.com/scala/scala/pull/9655/files#diff-1d7e8b347eefe51392356ff23a3a480317df5d42fc10d12336c344bd1407dc75L417